libflux/src/semantic/nodes.rs

```diff
@@ -58,6 +96,58 @@ pub enum Expression {
     Regexp(RegexpLit),
 }
 
+impl Expression {
+    fn infer(&mut self, env: Environment, f: &mut Fresher) -> Result {
+        match self {
+            Expression::Identifier(e) => e.infer(env, f),
+            Expression::Array(e) => e.infer(env, f),
+            Expression::Function(e) => e.infer(env, f),
+            Expression::Logical(e) => e.infer(env, f),
+            Expression::Object(e) => e.infer(env, f),
+            Expression::Member(e) => e.infer(env, f),
+            Expression::Index(e) => e.infer(env, f),
+            Expression::Binary(e) => e.infer(env, f),
+            Expression::Unary(e) => e.infer(env, f),
+            Expression::Call(e) => e.infer(env, f),
+            Expression::Conditional(e) => e.infer(env, f),
+            Expression::StringExpr(e) => e.infer(env, f),
+            Expression::Integer(lit) => lit.infer(env, f),
+            Expression::Float(lit) => lit.infer(env, f),
+            Expression::StringLit(lit) => lit.infer(env, f),
+            Expression::Duration(lit) => lit.infer(env, f),
+            Expression::Uint(lit) => lit.infer(env, f),
+            Expression::Boolean(lit) => lit.infer(env, f),
+            Expression::DateTime(lit) => lit.infer(env, f),
+            Expression::Regexp(lit) => lit.infer(env, f),
+        }
+    }
+}
+
+pub fn type_of(expr: &Expression) -> MonoType {
```

Comment:
I'd rather return an `&Monotype` and let the caller `clone` if necessary.

Reply:
You're right that makes more sense. I will update.